### PR TITLE
support tcp proxy independent without kernel enhanced

### PR DESCRIPTION
**Pull request by lec-bit**
A new compilation macro, ISKERNEL_ENHANCED, has been introduced to determine whether the current running environment is an enhanced kernel.

On an enhanced kernel, all capabilities of kmesh are supported. On a non-enhanced kernel, only the L4 proxy capability of kmesh is supported.

To implement these capabilities, the following modifications have been made to the code:

1、Code that cannot be compiled or run due to kernel dependencies is included using compilation macros. For Go code, file-level conditional compilation is used. A new file bpf_kmesh_l4.go has been added, which is derived from the bpf_kmesh.go file by removing the parts of the code that depend on the enhanced kernel. add bpf_kmesh_common.go，it is L4 common function.

2、The compilation script does not compile or load kmesh.ko.

3、Custom functions are used to replace the structures and bpf_helper functions provided by the enhanced kernel. 
              The structure bpf_mem_ptr and the functions bpf_strncmp, bpf_strncpy are replaced with custom implementations.

### CODE REVIEW

**Comment by lec-bit** [comment]
image

**Comment by lec-bit** [comment]
image

**Comment by lec-bit** [comment]
新增编译宏 ISKERNEL_ENHANCED,用于判断当前运行环境是否是增强过的内核
在增强过的内核上支持kmesh的全部能力，
在未增强过的内核上支持kmesh的 L4代理能力
为实现上述能力，代码部分做出以下修改：
1、将因内核依赖导致无法编译运行的代码使用编译宏包括起来，对于go代码采用文件级别的条件编译，新增bpf_kmesh_l4.go文件，该文件来自于bpf_kmesh.go文件，将其中依赖增强内核部分代码剔除后得到。
2、在编译脚本中不去编译、加载kmesh.ko
3、使用自定义函数替代依赖的增强内核提供的结构体和bpf_helper函数，
	结构体bpf_mem_ptr以及函数bpf_strncmp,bpf_strncpy

File: pkg/bpf/bpf_kmesh_l4.go [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+	Info BpfInfo
+	Link link.Link
+	bpf2go.KmeshCgroupSockObjects
+}
+
+type BpfKmesh struct {
+	SockConn BpfSockConn
+}
+
+func (sc *BpfSockConn) NewBpf(cfg *Config) error {
```
Comment (supercharge-xsy) [inline pkg/bpf/bpf_kmesh_l4.go:49]: BpfSockConn相关的实现建议抽取为一个单独的文件，L7/L4分别调用

File: pkg/bpf/bpf_kmesh_l4.go [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+	if err = sc.SockConn.NewBpf(cfg); err != nil {
+		return sc, err
+	}
+	return sc, nil
+}
+
+func setInnerMap(spec *ebpf.CollectionSpec) {
+	var (
+		InnerMapKeySize    uint32 = 4
+		InnerMapDataLength uint32 = 1300
```
[the remark this answers is not in the document]
Comment (supercharge-xsy) [reply pkg/bpf/bpf_kmesh_l4.go:83]: 使用cgo引用，C.XXX，保持一致

File: bpf/kmesh/cgroup_sock.c [hunk 3]
```diff
@@ -49,7 +49,7 @@ static inline int sock4_traffic_control(struct bpf_sock_addr *ctx)
 	}
 	BPF_LOG(DEBUG, KMESH, "bpf find listener addr=[%u:%u]\n", ctx->user_ip4, ctx->user_port);
 
-#if KMESH_ENABLE_HTTP
+#if ENHANCED_KERNEL
```
Comment (supercharge-xsy) [inline bpf/kmesh/cgroup_sock.c:52]: 刷新todo和配套的注释

File: build.sh [hunk 4]
```diff
@@ -1,6 +1,24 @@
...
+    if [ "$ENHANCED_KERNEL" == "enhanced" ]; then
+            sed -i '/tracepoint/s/\(.*\)generate/\/\/go:generate/' bpf/kmesh/bpf2go/bpf2go.go
+            sed -i '/sockops/s/\(.*\)generate/\/\/go:generate/' bpf/kmesh/bpf2go/bpf2go.go
+    else
+            sed -i '/tracepoint/s/\(.*\)generate/\/\/not go:generate/' bpf/kmesh/bpf2go/bpf2go.go
+            sed -i '/sockops/s/\(.*\)generate/\/\/not go:generate/' bpf/kmesh/bpf2go/bpf2go.go
+    fi
+}
+
+function is_enhanced_kernel() {
```
Comment (nlgwcy) [inline build.sh:14]: 函数名需要调整下，函数功能是根据helper判断当前是否为增强内核环境；建议改成：set_enhanced_kerenl_env....

**Pull request merged**